src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+    if (hashsetSize(s) == 0) return 0;
+    uint8_t h2 = highBits(hash);
+    int table;
+
+    /* Do some incremental rehashing. */
+    rehashStepOnReadIfNeeded(s);
+
+    /* Check rehashing destination table first, since it is newer and typically
+     * has less 'everfull' flagged buckets. Therefore it needs less probing for
+     * lookup. */
+    for (table = 1; table >= 0; table--) {
+        if (s->used[table] == 0) continue;
+        size_t mask = expToMask(s->bucketExp[table]);
+        size_t bucket_idx = hash & mask;
+        size_t start_bucket_idx = bucket_idx;
+        while (1) {
+            bucket *b = &s->tables[table][bucket_idx];
+            /* Find candidate elements with presence flag set and matching h2 hash. */
+            for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                if ((b->presence & (1 << pos)) && b->hashes[pos] == h2) {
+                    /* It's a candidate. */
+                    void *element = b->elements[pos];
+                    const void *elem_key = elementGetKey(s, element);
+                    if (compareKeys(s, key, elem_key) == 0) {
+                        /* It's a match. */
+                        if (pos_in_bucket) *pos_in_bucket = pos;
+                        if (table_index) *table_index = table;
+                        return b;
+                    }
+                }
+            }
+
+            /* Probe the next bucket? */
+            if (!b->everfull) break;
+            bucket_idx = nextCursor(bucket_idx, mask);
+            if (bucket_idx == start_bucket_idx) {
+                /* We probed the whole table. This should be extremely rare but
+                 * theoretically it can happen. */
+                break;
+            }
+        }
+    }
+    return NULL;
+}
+
+/* Find an empty position in the table for inserting an element with the given hash. */
+static bucket *findBucketForInsert(hashset *s, uint64_t hash, int *pos_in_bucket, int *table_index) {
+    int table = hashsetIsRehashing(s) ? 1 : 0;
+    assert(s->tables[table]);
+    size_t mask = expToMask(s->bucketExp[table]);
+    size_t bucket_idx = hash & mask;
+    while (1) {
+        bucket *b = &s->tables[table][bucket_idx];
+        for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+            if (b->presence & (1 << pos)) continue; /* busy */
+            if (pos_in_bucket) *pos_in_bucket = pos;
+            if (table_index) *table_index = table;
+            return b;
+        }
+        bucket_idx = nextCursor(bucket_idx, mask);
+    }
+}
+
+/* Encode bucket_index, pos_in_bucket, table_index into an opaque pointer. */
+static void *encodePositionInTable(size_t bucket_index, int pos_in_bucket, int table_index) {
+    uintptr_t encoded = bucket_index;
+    encoded <<= BITS_NEEDED_TO_STORE_POS_WITHIN_BUCKET;
+    encoded |= pos_in_bucket;
+    encoded <<= 1;
+    encoded |= table_index;
+    encoded++; /* Add one to make sure we don't return NULL. */
+    return (void *)encoded;
+}
```

Comment:
This isn't good for very large tables in 32-bit mode. We don't have enough bits.

I've considered just returning a pointer to the element's location in the bucket. It's possible to decode that too, back to bucket index and the other indexes.